### PR TITLE
Updating news and collections

### PR DESCRIPTION
**Pull request by danieltruong**
https://bcmines.atlassian.net/browse/NRPT-333
https://bcmines.atlassian.net/browse/FLCN-37
https://bcmines.atlassian.net/browse/NRPT-367
https://bcmines.atlassian.net/browse/NRPT-368
https://bcmines.atlassian.net/browse/NRPT-377
https://bcmines.atlassian.net/browse/NRPT-388

### CODE REVIEW

**Review by dhlevi — CHANGES_REQUESTED** [review]
That's a lot of work! Code-wise everything looks fine, just a couple of small comments.

Currently though I cannot create/edit collections or news items smoothly. Every attempt to save throws errors. Sometimes when I create a new collection, an error is thrown but the collection is created, sometimes not. Edits seem to always fail. 

Creating news items always throws an error, but seems to create the item successfully. Changes to the Title are ignored on create, but can be changed on update (description/url seems fine).

I'm going to debug a bit and see if I can spot the causes of some of the wonkiness. There are a bunch of tickets related to collection creating and updating as well so not sure entirely what should be covered by all your work here or by those tickets specifically... probably a bunch of overlap!

File: angular/projects/admin-nrpti/src/app/news/news-add-edit/news-add-edit.component.ts [hunk 1]
```diff
@@ -128,19 +128,20 @@ export class NewsAddEditComponent implements OnInit, OnDestroy {
...
+      if (!res || res.insertedCount !== 1) {
+        alert('Failed to create News Item.');
+      } else {
+        this.router.navigate(['news', this.record.system, res.insertedId, 'detail']);
+      }
     } else {
       // Update the news item.
-      this.factoryService.editNews(newsItem).subscribe(async res => {
+      const res = await this.factoryService.editNews(newsItem);
+      if (!res || res.modifiedCount !== 1) {
```
Comment (dhlevi) [inline angular/projects/admin-nrpti/src/app/news/news-add-edit/news-add-edit.component.ts:140]: should be:
`if (!res || res.nModified !== 1) {`

or alternatively res.ok !== 1

File: angular/projects/admin-nrpti/src/app/news/news-add-edit/news-add-edit.component.ts [hunk 2]
```diff
@@ -128,19 +128,20 @@ export class NewsAddEditComponent implements OnInit, OnDestroy {
...
-      this.factoryService.createNews(newsItem).subscribe(async res => {
-        // TODO: Need to check for res.length, as the return is typed object, but it's actually an array.
-        if (!res /*|| !res.length*/ || !res[0] || !res[0].length || !res[0][0]) {
-          alert('Failed to create News Item.');
-        } else {
-          this.router.navigate(['news', this.record.system, res[0][0].object._id, 'detail']);
-        }
-      });
+      const res = await this.factoryService.createNews(newsItem);
+      if (!res || res.insertedCount !== 1) {
```
Comment (dhlevi) [inline angular/projects/admin-nrpti/src/app/news/news-add-edit/news-add-edit.component.ts:132]: Should be:
`if (!res || res.ok !== 1) {`

Or, if you're sending back a response from the API, double check you're sending back the custom response and not the mongo response.

File: angular/projects/admin-nrpti/src/app/mines/mines-collections-add-edit/mines-collections-add-edit.component.ts [hunk 3]
```diff
@@ -405,34 +378,30 @@ export class MinesCollectionsAddEditComponent implements OnInit, OnDestroy {
...
     } else {
       collection['_master'] = this.route.snapshot.paramMap.get('mineId');
       collection['project'] = this.route.snapshot.paramMap.get('mineId');
 
-      this.factoryService.createCollection(collection).subscribe(async (res: any) => {
-        this.recordUtils.parseResForErrors(res);
-
-        const createdCollection = res && res.length && res[0] && res[0].length && res[0][0] && res[0][0].object;
+      const res = await this.factoryService.createCollection(collection);
+      if (!res || res.insertedCount !== 1) {
```
Comment (dhlevi) [inline angular/projects/admin-nrpti/src/app/mines/mines-collections-add-edit/mines-collections-add-edit.component.ts:393]: Should be: 
`if (!res || res.ok !== 1)`

File: angular/projects/admin-nrpti/src/app/mines/mines-collections-add-edit/mines-collections-add-edit.component.ts [hunk 4]
```diff
@@ -405,34 +378,30 @@ export class MinesCollectionsAddEditComponent implements OnInit, OnDestroy {
...
       collection['_id'] = this.collection._id;
 
-      this.factoryService.editCollection(collection).subscribe(async res => {
-        this.recordUtils.parseResForErrors(res);
-
-        // update the records collectionId
-        this.updateRecordsCollectionId(this.collection._id);
-
+      const res = await this.factoryService.editCollection(collection);
+      if (!res || res.modifiedCount !== 1) {
```
Comment (dhlevi) [inline angular/projects/admin-nrpti/src/app/mines/mines-collections-add-edit/mines-collections-add-edit.component.ts:382]: Should be: 
`if (!res || res.nModified !== 1)` or `if (!res || res.ok !== 1)`

File: api/src/controllers/news-controller.js [hunk 5]
```diff
@@ -40,47 +44,79 @@ exports.protectedPut = async function (args, res, next) {
...
+  news._schemaName = 'ActivityLNG';
+
+  incomingObj._epicProjectId &&
+    ObjectID.isValid(incomingObj._epicProjectId) &&
+    (news._epicProjectId = new ObjectID(incomingObj._epicProjectId));
+
+  incomingObj.description && (news.description = incomingObj.description);
+  incomingObj.projectName && (news.projectName = incomingObj.projectName);
+  incomingObj.type && (news.type = incomingObj.type);
+  incomingObj.type && (news.type = incomingObj.type);
```
Comment (dhlevi) [inline api/src/controllers/news-controller.js:104]: Looks like a duped line. I'm guessing it should be:
`incomingObj.title && (news.type = incomingObj.title);`

Comment (danieltruong) [reply api/src/controllers/news-controller.js]: 3a32d6f8c9e96a7ef201e56a13995a6357f2498e

File: angular/projects/admin-nrpti/src/app/mines/mines-collections-add-edit/mines-collections-add-edit.component.ts (hunk 4 above)
Comment (danieltruong) [reply angular/projects/admin-nrpti/src/app/mines/mines-collections-add-edit/mines-collections-add-edit.component.ts]: 3a32d6f8c9e96a7ef201e56a13995a6357f2498e

File: angular/projects/admin-nrpti/src/app/mines/mines-collections-add-edit/mines-collections-add-edit.component.ts (hunk 3 above)
Comment (danieltruong) [reply angular/projects/admin-nrpti/src/app/mines/mines-collections-add-edit/mines-collections-add-edit.component.ts]: 3a32d6f8c9e96a7ef201e56a13995a6357f2498e

File: angular/projects/admin-nrpti/src/app/news/news-add-edit/news-add-edit.component.ts (hunk 2 above)
Comment (danieltruong) [reply angular/projects/admin-nrpti/src/app/news/news-add-edit/news-add-edit.component.ts]: 3a32d6f8c9e96a7ef201e56a13995a6357f2498e

File: angular/projects/admin-nrpti/src/app/news/news-add-edit/news-add-edit.component.ts (hunk 1 above)
Comment (danieltruong) [reply angular/projects/admin-nrpti/src/app/news/news-add-edit/news-add-edit.component.ts]: 3a32d6f8c9e96a7ef201e56a13995a6357f2498e

File: api/src/controllers/news-controller.js [hunk 6]
```diff
@@ -40,47 +44,79 @@ exports.protectedPut = async function (args, res, next) {
...
+
+  let obj = null;
   try {
-    await Put.updateById(newsId, news);
+    obj = await Put.updateById(newsId, news);
   } catch (error) {
     defaultLog.info(`protectedPut - error updating news: ${newsId}`);
     defaultLog.debug(error);
-    return queryActions.sendResponse(res, 400, {});
+    console.log(news);
```
[the remark this answers is not in the document]
Comment (danieltruong) [reply api/src/controllers/news-controller.js]: 3a32d6f8c9e96a7ef201e56a13995a6357f2498e

File: api/src/controllers/record-controller.js [hunk 7]
```diff
@@ -264,6 +259,7 @@ exports.protectedDelete = async function (args, res, next) {
   } catch (error) {
     defaultLog.info(`protectedDelete - error deleting record: ${recordId}`);
     defaultLog.debug(error);
+    console.log(error);
```
[the remark this answers is not in the document]
Comment (danieltruong) [reply api/src/controllers/record-controller.js]: 3a32d6f8c9e96a7ef201e56a13995a6357f2498e

File: api/src/controllers/record-controller.js [hunk 8]
```diff
@@ -248,6 +242,7 @@ exports.protectedDelete = async function (args, res, next) {
   }
 
   try {
+    console.log(RecordTypeEnum.BCMI_SCHEMA_NAMES);
```
[the remark this answers is not in the document]
Comment (danieltruong) [reply api/src/controllers/record-controller.js]: 3a32d6f8c9e96a7ef201e56a13995a6357f2498e

File: angular/projects/admin-nrpti/src/app/services/news.service.ts [hunk 9]
```diff
@@ -5,7 +5,28 @@ import { ApiService } from './api.service';
...
+    if (!news) {
+      throw Error('NewsService - createNews - missing required news param');
+    }
+
+    const queryString = 'news/';
+    return this.http.post<any>(`${this.apiService.pathAPI}/${queryString}`, news, {}).toPromise();
+  }
+
+  public editNews(news: any): Promise<any> {
+    if (!news) {
```
[the remark this answers is not in the document]
Comment (danieltruong) [reply angular/projects/admin-nrpti/src/app/services/news.service.ts]: 3a32d6f8c9e96a7ef201e56a13995a6357f2498e

File: api/src/controllers/collection-controller.js [hunk 10]
```diff
@@ -106,4 +252,29 @@ exports.protectedDelete = async function (args, res, next) {
...
   next();
+}
+
+const checkRecordExistsInCollection = async function (records, collectionId, editing = false) {
+  const db = mongodb.connection.db(process.env.MONGODB_DATABASE || 'nrpti-dev');
+  const collectionDB = db.collection('nrpti');
+  let promises = [];
+  for (const record of records) {
+    // does this record exit in any other collection?
+    const collectionCount = await collectionDB.countDocuments({ _schemaName: RECORD_TYPE.CollectionBCMI._schemaName, records: { $elemMatch: { $eq: new ObjectID(record) } } });
```
Comment (mtCarto) [inline api/src/controllers/collection-controller.js:263]: trying to update an newly created and edited collection (note my records had no attached documents) the api crashes due to TypeError: collectionDB.countDocuments is not a function

File: angular/projects/admin-nrpti/src/app/mines/mines-collections-add-edit/mines-collections-add-edit.component.ts [hunk 11]
```diff
@@ -405,34 +378,30 @@ export class MinesCollectionsAddEditComponent implements OnInit, OnDestroy {
...
 
-      this.factoryService.createCollection(collection).subscribe(async (res: any) => {
-        this.recordUtils.parseResForErrors(res);
-
-        const createdCollection = res && res.length && res[0] && res[0].length && res[0][0] && res[0][0].object;
+      const res = await this.factoryService.createCollection(collection);
+      if (!res || res.ok !== 1) {
+        alert('Failed to create collection.');
+      } else {
+        const createdCollection = res.ops[0];
```
Comment (mtCarto) [inline angular/projects/admin-nrpti/src/app/mines/mines-collections-add-edit/mines-collections-add-edit.component.ts:396]: creating collection and saving (with no added records) gets stuck on the spinner and throws res.ops is undefined
but the collection does get created

**Review by mtCarto — CHANGES_REQUESTED** [review]

**Comment by mtCarto** [comment]
Creating a news item works, but throws an error and doesn't return to list page. 
Editing and deleting seem to be working

**Comment by dhlevi** [comment]
Ah, It looks like it's the returns from the API. You've got something like:

```
obj = await Put.updateById(collectionId, updateObj);
queryActions.sendResponse(res, 200, obj);
```

On the put/posts, so it's returning the result from the mongoose process, but it looks like you meant to send back a different object. Likely the issues that we're seeing can be fixed by either changing the API up to use the custom object that was probably there originally, or just updating the angular side to work with the new results coming back.

Explains why they all seem to work, but throw errors on the angular side!

**Comment by danieltruong** [comment]
Turns out I was using MongoDB V4 functions which makes sense that it works for me but not you guys. @dhlevi @mtCarto

File: api/src/controllers/collection-controller.js (hunk 10 above)
Comment (danieltruong) [reply api/src/controllers/collection-controller.js]: c4001780a78be7845c544c94e828bec4fb8b1900

**Comment by danieltruong** [comment]
> Ah, It looks like it's the returns from the API. You've got something like:
> 
> ```
> obj = await Put.updateById(collectionId, updateObj);
> queryActions.sendResponse(res, 200, obj);
> ```
> 
> On the put/posts, so it's returning the result from the mongoose process, but it looks like you meant to send back a different object. Likely the issues that we're seeing can be fixed by either changing the API up to use the custom object that was probably there originally, or just updating the angular side to work with the new results coming back.
> 
> Explains why they all seem to work, but throw errors on the angular side!

I can't seem to get the same errors you guys are getting. I know I've changed what the api sends to the angular side, but I've also changed the angular side to account for this. What errors are being thrown in the front end?

**Review by mtCarto — APPROVED** [review]
Looks good, error free!

**Review by dhlevi — APPROVED** [review]
A lot of weird issues but looks like it's behaving itself now. Changes look good, collection and news API side are functioning as expected!

**Pull request merged**